src/OpenTelemetry/Logs/OpenTelemetryLogger.cs

```diff
@@ -49,18 +49,16 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
             var processor = this.provider.Processor;
             if (processor != null)
             {
-                var options = this.provider.Options;
-
                 var record = new LogRecord(
-                    options.IncludeScopes ? this.ScopeProvider : null,
+                    this.provider.IncludeScopes ? this.ScopeProvider : null,
                     DateTime.UtcNow,
                     this.categoryName,
                     logLevel,
                     eventId,
-                    options.IncludeFormattedMessage ? formatter?.Invoke(state, exception) : null,
-                    options.ParseStateValues ? null : state,
+                    this.provider.IncludeFormattedMessage ? formatter?.Invoke(state, exception) : null,
```

Comment:
The `this.provider` is used multiple times here which caused multiple `ldfld` IL to be generated.

Need to 1) measure the perf given this is on hot path 2) reduce the cost

Reply:
It should be easy to avoid this by having these fields in the Logger class itself.

Reply:
I can see several options. In general we need to make sure that any modifications on the hot path are measured against perf impact (and if the noise level is high, we need to find better way to measure instead of drifting to poor perf).

Reply:
@reyang @cijothomas do you have any advice for running the Benchmarks?
I'm doing back-to-back runs and getting different results.

Reply:
Where is the benchmark code?

Reply:
I found BenchmarkDotNet's [best practices](https://benchmarkdotnet.org/articles/guides/good-practices.html).
I shut down most of my programs and got much more consistent results.

I was following the instructions [here](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/test/Benchmarks) and running the tests in [LogBenchmarks.cs](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Logs/LogBenchmarks.cs)




### main branch
```
|                                 Method |       Mean |     Error |     StdDev |  Gen 0 | Allocated |
|--------------------------------------- |-----------:|----------:|-----------:|-------:|----------:|
|                             NoListener |  65.527 ns | 1.3479 ns |  2.7836 ns | 0.0101 |      64 B |
|   NoListenerWithLoggerMessageGenerator |   5.247 ns | 0.1241 ns |  0.1740 ns |      - |         - |
|                           OneProcessor | 172.858 ns | 3.3029 ns |  4.5211 ns | 0.0370 |     232 B |
| OneProcessorWithLoggerMessageGenerator | 138.863 ns | 2.8119 ns |  6.2309 ns | 0.0267 |     168 B |
|                          TwoProcessors | 186.247 ns | 3.7727 ns | 10.7026 ns | 0.0370 |     232 B |
|                        ThreeProcessors | 183.196 ns | 3.6586 ns |  8.9747 ns | 0.0370 |     232 B |
```


### WITH this.options `ldfld`
```
|                                 Method |       Mean |     Error |     StdDev |  Gen 0 | Allocated |
|--------------------------------------- |-----------:|----------:|-----------:|-------:|----------:|
|                             NoListener |  63.320 ns | 1.1267 ns |  1.5794 ns | 0.0101 |      64 B |
|   NoListenerWithLoggerMessageGenerator |   4.862 ns | 0.0757 ns |  0.0708 ns |      - |         - |
|                           OneProcessor | 159.104 ns | 3.1748 ns |  5.5605 ns | 0.0370 |     232 B |
| OneProcessorWithLoggerMessageGenerator | 135.320 ns | 2.6625 ns |  5.0657 ns | 0.0267 |     168 B |
|                          TwoProcessors | 160.163 ns | 2.8330 ns |  3.6837 ns | 0.0370 |     232 B |
|                        ThreeProcessors | 196.035 ns | 4.1022 ns | 11.7039 ns | 0.0370 |     232 B |
```

### WITH copying fields to the Logger class
```
|                                 Method |       Mean |     Error |    StdDev |  Gen 0 | Allocated |
|--------------------------------------- |-----------:|----------:|----------:|-------:|----------:|
|                             NoListener |  59.460 ns | 1.1879 ns | 1.6653 ns | 0.0101 |      64 B |
|   NoListenerWithLoggerMessageGenerator |   4.888 ns | 0.1231 ns | 0.1152 ns |      - |         - |
|                           OneProcessor | 156.369 ns | 2.6028 ns | 2.1735 ns | 0.0370 |     232 B |
| OneProcessorWithLoggerMessageGenerator | 121.974 ns | 2.0873 ns | 1.9524 ns | 0.0267 |     168 B |
|                          TwoProcessors | 161.967 ns | 3.0390 ns | 2.5377 ns | 0.0370 |     232 B |
|                        ThreeProcessors | 156.933 ns | 1.7382 ns | 1.5408 ns | 0.0370 |     232 B |
```

Reply:
Thanks, I'll unblock this PR. Please include the following things in the PR description:
1. what are the possible options.
2. what's the perf analysis for each options.
3. why would you go with a certain option, what's the perf difference before/after.

Reply:
Here goes a good example #2805

Reply:
@reyang @cijothomas I updated the description with more targeted benchmarks.
The results are inconclusive, I'm seeing a negligible difference between implementations.
Is this good enough to proceed?

Reply:
Where is the benchmark code? Consider following what's done in https://github.com/dotnet/runtime/issues/67207:

![image](https://user-images.githubusercontent.com/17327289/160470617-ebcdca64-d411-41dd-83bd-2410af72744e.png)

Reply:
> The results are inconclusive, I'm seeing a negligible difference between implementations.

This is essentially telling us the measurement was performed in the incorrect place or the tool is insufficient for this kind of task.

Reply:
updated. 

side note, currently looking for a tool to inspect the IL code

Reply:
After reviewing the IL, I can see that Option 1 introduced several extra `ldfld` which was Reiley's concern.

The Original implementation is nearly identical to Options 2 & 3. 

```
ORIGINAL
	IL_002d: callvirt instance bool OpenTelemetry.Logs.OpenTelemetryLoggerOptions::get_IncludeScopes()

OPTION 1: LDFLD
	IL_0024: ldfld class OpenTelemetry.Logs.OpenTelemetryLoggerProvider2 OpenTelemetry.Logs.OpenTelemetryLogger_LDFLD::provider
	IL_0029: ldfld bool OpenTelemetry.Logs.OpenTelemetryLoggerProvider2::IncludeScopes

OPTION 2: LOCAL COPY
	IL_0021: ldfld bool OpenTelemetry.Logs.OpenTelemetryLogger_LocalCopy::IncludeScopes

OPTION 3: VAR PROVIDER
	IL_0023: ldfld bool OpenTelemetry.Logs.OpenTelemetryLoggerProvider2::IncludeScopes
```

I believe the best approach is to create a local `var provider` from which all other fields/properties can be accessed.

```csharp
            var provider = this.provider;
            var processor = provider.Processor;
            if (processor != null)
            {
                var record = new LogRecord(
                    provider.IncludeScopes ? this.ScopeProvider : null,
                    DateTime.UtcNow,
                    this.categoryName,
                    logLevel,
                    eventId,
                    provider.IncludeFormattedMessage ? formatter?.Invoke(state, exception) : null,
                    provider.ParseStateValues ? null : state,
                    exception,
                    provider.ParseStateValues ? this.ParseState(state) : null);

                processor.OnEnd(record);

                record.ScopeProvider = null;
            }
```